### DISCUSSION TITLE
File format specification

### DISCUSSION THREAD

**Discussion by peter88213** [vkbo/novelWriter#1011] [Q&A] [answered]
Is there a specification for the .nwx project file format that says something about the order of content item entries?
With my [yWriter converter](https://peter88213.github.io/yw2nw/changelog), I first built the project tree based on the parent entries and then serialized it again. In the last version, I abandoned that and just used the flat XML structure to obtain the document structure. 
As it looks, novelWriter writes the parts, chapters and scenes of the document in the right order and puts "outtakes" and "trash" at the end. Thus, the "Parent" information for each node of the project tree is basically redundant. Or not so? Can one rely on the project tree being properly serialized in the .nwx project file?

**Comment by vkbo** [comment, marked as answer]
The order of the content in the XML is the linear order of items in the project tree from top to bottom. Linear meaning it ignores the parent/child hierarchy (column number) and follows the row number. You can rely on this being the case when novelWriter writes the XML. However, the parent reference is needed in order to re-create the tree structure when the project is loaded, otherwise it would end up as a flat list. The order attribute is presently redundant though. It is mostly written to the XML for debug purposes.

The algorithm that loads the project tree is able to handle cases where the content entries in the file have been re-ordered or appear in the wrong order from a hierarchical point of view. It will detect, for instance, if a child item is loaded before its parent item, in which case it will postpone the loading of that item and try again later. So the loading is robust in terms of parent/child relationship, but a re-ordered XML will still mess up the order of sibling items at any given level. Postponing the loading of an item will also put it in the wrong location after the tree is built.

Since the intended order is saved to the XML, I could write code that reorders the final tree based on the order attribute in the XML, but since there is no scenario internally in the XML handling that could produce such an outcome as it stands. Therefore I never prioritised this. Of course, when the file is generated externally such a scenario is possible, so if needed I can look into this. In that case, the order of entries in the XML would in principle no longer matter at all.

The early versions of novelWriter ran on Python 3.4 and up, and prior to 3.6 the entries of a dictionary were ordered by their hash key. Since Python 3.6 the order is effectively the same as they were added, although this isn't guaranteed until Python 3.7. Due to the history of Python version support, the project content is stored separately (in a dict) from the project order (in a list) and there are the mentioned checks on loading that makes sure everything is ok before it is loaded into the project tree widget. From next release I'm dropping Python 3.6, so now it's actually possible to redesign this process, but I will not change the XML format in any case.

For reference, the function that serves the entries to the tree widget during load time can be found here: https://github.com/vkbo/novelWriter/blob/7f43584ed535a421887916e79b8074bb725850cc/novelwriter/core/project.py#L1143-L1183

**Reply by peter88213** [reply]
Thank you for clarifying this. 

> Due to the history of Python version support, the project content is stored separately (in a dict) from the project order (in a list) 

This is also [how I implemented the yWriter metamodel](https://peter88213.github.io/PyWriter/lib/model), although I am aware that since (C)Python 3.6 dictionaries are ordered by design. But when restructuring or merging projects, the lists seem a bit handier to me.

But essential for me is the convention that the order in the XML file matters. This allows for much more compact code for reading the file. 
Then I can easily build the document structure from the heading hierarchy. 
When creating a _.nwx_ file, the correct sequence of items is automatically generated. Of course, I also create the "parent" attributes, but this again requires additional effort.

If there should ever be an official specification of your file format, it would definitely be beneficial to mention the aspect of order in the XML structure.  

Cheers, 
Peter

**Reply by vkbo** [reply]
Sure, I can add a file spec to the technical section of the docs.

**Reply by vkbo** [reply]
I've had the file format spec document for a while now, but I added it to the documentation today. It's available from the [Project Format Changes](https://novelwriter.readthedocs.io/en/latest/usage_projectformat.html) page. It is available as a [PDF](https://novelwriter.readthedocs.io/en/latest/_static/fileformatspec15.pdf). The raw document is in the [repo](https://github.com/vkbo/novelWriter/tree/main/docs).

In the code I've separated out the XML reading and writing into a separate module. The XML writer class is available here: https://github.com/vkbo/novelWriter/blob/ecff40dadeb3aad1be4aceceaf9396f87170fb5b/novelwriter/core/projectxml.py#L475-L485

It should make it easier to re-implement in various tool. I've modified the format to be a lot simpler in general, as is evident from the class.

**Comment by peter88213** [comment]
I'm looking at the new file format and am considering updating my [yWriter file converter](https://peter88213.github.io/yw2nw/). 
Your file format specification as PDF is good work. However, for version comparison, a DTD would be very helpful. Is there something like that for NovelWriter?

**Reply by vkbo** [reply]
No, there's no DTD. I have considered adding a schema, but since I neither use namespaces nor intend the XML format to be used as an interchange format, it didn't really feel natural to do this. The XML parser code validates the XML file on the fly according to the version number.

**Reply by peter88213** [reply]
I see. However, creating a DTD doesn't take too much work, and you have a formal specification that can be used for comparison during versioning (which, if I understand it correctly, is also a goal of the novelWriter format itself).

**Reply by vkbo** [reply]
DTDs are somewhat outdated and replaced by schemas, so I would still make a schema rather than a DTD. Even if I don't have namespaces (yet). However, they don't really help me to maintain the code and the parsing, so the only value is for converter tools. Perhaps.

The XML reader class is backward compatible though, and will convert any released format on the fly to the latest one. Converting to a newer format in a conversion tool is only strictly beneficial if it adds new features. Aside from being more compact, most of the changes in 1.4 and 1.5 are intended for caching data between sessions for new internal functionality in novelWriter. They don't add any actual new information that isn't already there in 1.3. The only exception is the splitting of `status` and `import` from 1.3 to 1.4. I also impose some restrictions that makes the format more compliant with the XML standard, but that too is handled both on read and write.

**Reply by peter88213** [reply]
Yes, I just realized that my code for reading the 1.4 file format mostly fits for 1.5 as well. What still needs to be changed is looking up the status and importance via the keys.

I guess it's also not a problem if my converter continues to write only the 1.3 format, is it?

**Reply by vkbo** [reply]
Yeah, the change to using keys for status and importance is the main one affecting a converter. As long as you generate valid and unique keys, you can just build a map in the converter. That's what the internal converter for 1.3 to 1.4+ does anyway. There is nothing that prevents you from generating sequential keys either. I do this in the test suite, as it makes the keys predictable for testing. So `s000001`, `s000002` etc for status and `i000001`, `i000002` etc for importance works just fine. The reason the code uses random numbers is that it eliminates the need for remembering which number is the next one in the sequence.

To answer your second question, 1.3 format is still fully supported, yes. If you haven't looked at it already the parsing code is now all located in one file, and is written in a way that avoids bloated parsing code when changes are made. As such, there is not really any need to remove support for deprecated XML nodes. The parser is generally unopinionated and optimistic, and saves whatever correctly labelled data it can find. There are very few nodes and attributes in the spec that are required. https://github.com/vkbo/novelWriter/blob/main/novelwriter/core/projectxml.py

A lot of the new attributes added in 1.4 and 1.5 are re-generated on the fly if they are missing. Internally, there is a closer connection between the index and the project storage than there used to be (some info used to be duplicated between the two). This is because the project tree now also depends on indexed data to pick icons and other bits, which didn't use to be the case. But since the index is automatically rebuilt after conversion with your code, which we dealt with a while back, all of the missing settings are restored too. This includes all attributes in the `meta` node, and also the `root` attribute in the `item` node.

Part of the change is also that I've standardised how data is stored. I have one way if storing Python dictionary entries (the `entry` nodes) and moved everything that is pure meta data into attributes for the relevant section nodes. Everything is now consistent, so for the foreseeable future, changes to the XML file should be additive. I only bump versions when there are deprecations.

Also, the XML now uses "yes" and "no" for logical flags, as I wanted it to be less Python literal, but it supports the old logicals as well, so you don't need to care about this.

https://github.com/vkbo/novelWriter/blob/66cc3777dc5072b5139013bfcd44e4edf5b83336/novelwriter/common.py#L90-L97

**Reply by peter88213** [reply]
Yes, it is beneficial to generate keys in a reproducible manner for several reasons. I solved the problem by making a class generate and store them, see: https://github.com/peter88213/yw2nw/blob/main/src/yw2nwlib/handles.py
This code is a rather old one; it's more elegant to let the class inherit from the list type.

**Reply by vkbo** [reply]
The current handle generator code in novelWriter is a one-liner:

```python
handle = f"{random.getrandbits(52):013x}"
```

You can replace it with a sequential generator by replacing `random.getrandbits(52)` with an incrementing integer variable which could of course just be the size of the list or dictionary they are stored in, which will always be one larger than the last used integer value.

The code is located here:

https://github.com/vkbo/novelWriter/blob/66cc3777dc5072b5139013bfcd44e4edf5b83336/novelwriter/core/tree.py#L436-L446

The chance of a duplicate handle is very slim, one in 4500 trillion, but the code is a remnant from when the handle was generated using the current timestamp, which failed often on Windows due to bad tick resolution. Still, these are pseudo-random numbers so the real chance may be higher. I've decided to leave the collision checker and ignore this detail.

The status and importance keys are generated the same way in, using a `self._prefix` of either `s` or `i`:

https://github.com/vkbo/novelWriter/blob/66cc3777dc5072b5139013bfcd44e4edf5b83336/novelwriter/core/status.py#L236-L246

**Reply by vkbo** [reply]
On another note, is the converter tool functional now so that I can link to it from the documentation?

I am planning on adding a HowTo section one I have enough "hows" to write one.

**Reply by peter88213** [reply]
At the moment, the specification is as follows:

- **Read and write** file format version 1.3 created with **novelWriter 1.5 and 1.6**
- **Read** file format version 1.4 created with **novelWriter 1.7**

For file format v1.5 support, I opened an issue and started a branch. However, I cannot make no statement as to when the release will be ready. Actually, I wanted to interpose the topic today, but it looks like the weekend is already over and I have to turn away from the computer screen now.

**Reply by vkbo** [reply]
As long as there is a text that describes which versions are supported, there is no problem adding it to the docs. But as the weekend is indeed over, I will not do anything on this today either.

As a sidenote, there was never a release 1.7, hence there is probably no 1.4 format projects out there save for a few users who tested the 2.0 RC 1 pre-release. I would just ignore 1.4 altogether. Anyone using it should anyway update to the full 2.0 release which uses project format 1.5.

**Reply by peter88213** [reply]
That's okay with me. 
See: https://github.com/peter88213/yw2nw/issues/2

**Reply by peter88213** [reply]
Meanwhile, I implemented full v 1.5 support, and dropped v 1.3. With that, I'm ending the development of this tool unless serious bugs are reported. After all, I have no idea if there is anyone out there who uses this.
